File: IDI Open/solutions/2012/candy/candy_rs.c
/*
	solution by ruben spaans
	calculate 2^N - sum(i=0..c-1) dp[i] using dp
*/

#include <stdio.h>
#include <string.h>
#define MOD 65537
int dp[10000];
int n,c;
int a[200];

int main() {
	int T,p2,i,j,maks,v;
	scanf("%d",&T);
	while(T--) {
		scanf("%d %d",&n,&c);
		for(p2=1,i=0;i<n;i++) {
			scanf("%d",&a[i]);
			p2=p2*2%65537;
		}
		memset(dp,0,c*sizeof(int));
		dp[0]=1;
		for(maks=i=0;i<n;i++) {
			for(j=maks>=c?c-1:maks;j>=0;j--) if((v=dp[j])) {
				if(j+a[i]<c) dp[j+a[i]]=(dp[j+a[i]]+v)%MOD;
			}
			maks+=a[i];
		}
		for(v=i=0;i<c;i++) v=(v+dp[i])%MOD;
		printf("%d\n",(p2+MOD-v)%MOD);
	}
	return 0;
}

File: IDI Open/solutions/2012/candy/candy_rs.java
/*
	solution by ruben spaans, regular dp
*/

import java.io.*;

public class candy_rs {
	public static void main(String[] a) throws Exception {
		new candy_rs().go();
	}
	int MOD=65537;
	void go() throws Exception {
		BufferedReader in=new BufferedReader(new InputStreamReader(System.in));
		int T=Integer.parseInt(in.readLine());
		while(T-->0) {
			int N,C;
			String s[]=in.readLine().split(" ");
			N=Integer.parseInt(s[0]);
			C=Integer.parseInt(s[1]);
			int a[]=new int[N];
			int sum=0;
			s=in.readLine().split(" ");
			for(int i=0;i<N;i++) {
				a[i]=Integer.parseInt(s[i]);
				sum+=a[i];
			}
			int prev[],cur[]=new int[sum+1];
			cur[0]=1;
			for(int i=0;i<N;i++) {
				prev=cur;
				cur=new int[sum+1];
				for(int j=0;j<=sum;j++) if(prev[j]>0) {
					cur[j]=(cur[j]+prev[j])%MOD;
					cur[j+a[i]]=(cur[j+a[i]]+prev[j])%MOD;
				}
			}
			int ans=0;
			while(C<=sum) ans=(ans+cur[C++])%MOD;
			System.out.println(ans);
		}
	}
}
